### PR TITLE
fix: PATCH method bug

### LINKED ISSUE #570: [refactor][bug] PATCH method bug
PATCH method doesn't support subpath and embedded struct.

example:

```
$ curl http://127.0.0.1:8080/apisix/admin/routes/1/upstream/nodes -X PATCH -i -d '
{
    "39.97.63.200:80": 1
}'
```

### PR DESCRIPTION
**Pull request by nic-chen**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance

- Related issues
close #570 

___
### Bugfix
- Description

The old version uses the data after binding the structure as the data of the patch, and there will be a problem of 0 value after binding. 
And the old patch library has some bugs.

- How to fix?

Use raw data as patch data
Replace the patch library

### CODE REVIEW

File: api/internal/utils/json_patch.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+		out, err = obj.Apply(doc)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return out, nil
+}
+
+func MakePatch(obj interface{}, subPath string, reqBody []byte) ([]byte, error) {
```
Comment (liuxiran) [inline api/internal/utils/json_patch.go:58]: how about use `MergePatch` instead of `MakePatch`, `MakePatch` is easy to misunderstand as makediff

**Review by membphis — CHANGES_REQUESTED** [review]
need some E2E test cases

**Comment by nic-chen** [comment]
> need some E2E test cases

we already have E2E test cases for it:
https://github.com/apache/apisix-dashboard/pull/1005/files#diff-c2b7edfd95881ffeb7a02870b56d8c228b7b1f0f83d78be56f4462091bf76dc7R115-R174

File: api/internal/utils/json_patch.go (hunk 1 above)
Comment (nic-chen) [reply api/internal/utils/json_patch.go:58]: ok

File: api/internal/handler/ssl/ssl.go [hunk 2]
```diff
@@ -62,13 +62,14 @@ func (h *Handler) ApplyRoute(r *gin.Engine) {
...
 		wrapper.InputType(reflect.TypeOf(UpdateInput{}))))
-	r.PATCH("/apisix/admin/ssl/:id", wgin.Wraps(h.Patch,
-		wrapper.InputType(reflect.TypeOf(UpdateInput{}))))
 	r.DELETE("/apisix/admin/ssl/:ids", wgin.Wraps(h.BatchDelete,
 		wrapper.InputType(reflect.TypeOf(BatchDelete{}))))
 	r.POST("/apisix/admin/check_ssl_cert", wgin.Wraps(h.Validate,
 		wrapper.InputType(reflect.TypeOf(entity.SSL{}))))
 
+	r.PATCH("/apisix/admin/ssl/:id", consts.ErrorWrapper(Patch))
+	r.PATCH("/apisix/admin/ssl/:id/*path", consts.ErrorWrapper(Patch))
```
Comment (juzhiyuan) [inline api/internal/handler/ssl/ssl.go:71]: *path means any path?

File: api/internal/utils/json_patch_test.go [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+	return buf.String()
+}
+
+func TestMergeJson(t *testing.T) {
+	Cases := []struct {
+		doc, patch, result, desc string
+	}{
+		{
+			desc: `simple merge`,
```
Comment (juzhiyuan) [inline api/internal/utils/json_patch_test.go:47]: image

File: api/internal/handler/ssl/ssl.go (hunk 2 above)
Comment (nic-chen) [reply api/internal/handler/ssl/ssl.go:71]: * means any, `path` is a variable name for it.

usage of gin framework

**Review by membphis — APPROVED** [review]

File: api/internal/utils/json_patch_test.go [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+                                        },{
+                                                "host": "39.97.63.217",
+                                                "port": 80,
+                                                "weight" : 1
+                                        }]
+                                }
+                        }`,
+		},
+		{
+			desc: "patch field that not exists",
```
Comment (tokers) [inline api/internal/utils/json_patch_test.go:172]: not exist => non existent

File: api/internal/utils/json_patch_test.go [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+func formatJSON(j string) string {
+	buf := new(bytes.Buffer)
+
+	json.Indent(buf, []byte(j), "", "  ")
+
+	return buf.String()
+}
+
+func TestMergeJson(t *testing.T) {
+	Cases := []struct {
```
Comment (tokers) [inline api/internal/utils/json_patch_test.go:43]: use small camel case.

File: api/internal/handler/ssl/ssl.go [hunk 6]
```diff
@@ -182,40 +183,29 @@ func (h *Handler) Update(c droplet.Context) (interface{}, error) {
...
-			Operations: []jsonpatch.PatchOperation{
-				{Op: jsonpatch.Replace, Path: subPath, Value: c.Input()},
-			},
-		}
-	} else {
-		patch, err = jsonpatch.MakePatch(stored, input.SSL)
-		if err != nil {
-			return handler.SpecCodeResponse(err), err
-		}
+	res, err := utils.MergePatch(stored, subPath, reqBody)
```
[the remark this answers is not in the document]
Comment (nic-chen) [reply api/internal/handler/ssl/ssl.go:197]: @tokers  we have schema verify before create or update.

File: api/internal/utils/json_patch_test.go (hunk 4 above)
Comment (nic-chen) [reply api/internal/utils/json_patch_test.go]: fixed

File: api/internal/utils/json_patch_test.go (hunk 5 above)
Comment (nic-chen) [reply api/internal/utils/json_patch_test.go]: fixed

**Comment by juzhiyuan** [comment]
hi @nic-chen, please merge this PR[1] first, that PR fixes a bug when deploying to Azure.

[1] https://github.com/apache/apisix-dashboard/pull/1018

File: api/internal/handler/ssl/ssl.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (tokers) [reply api/internal/handler/ssl/ssl.go:197]: OK

File: api/internal/utils/json_patch_test.go (hunk 3 above)
Comment (nic-chen) [reply api/internal/utils/json_patch_test.go]: fixed

**Review by juzhiyuan — APPROVED** [review]

**Pull request merged**